### PR TITLE
Adjust shutil.move stubs to allow bytes/Pathlike[bytes] args

### LINKED ISSUE #6831: shutil.move stub does not account for bytes paths
Just encountered this.

Consider this valid script:
```python
import os
from pathlib import Path
import shutil
from typing import AnyStr


class DummyPathLike(os.PathLike[AnyStr]):
    """Since pathlib.Path can't accept bytes itself."""

    def __init__(self, path: AnyStr) -> None:
        self._path: AnyStr = path

    def __fspath__(self) -> AnyStr:
        """Return the file system path representation of the object."""
        return self._path


Path("test1.txt").touch()

# Both str
shutil.move("test1.txt", "test2.txt")

# Both bytes
shutil.move(b"test2.txt", b"test1.txt")

# Both Pathlike[str]
shutil.move(Path("test1.txt"), Path("test2.txt"))

# Both Pathlike[bytes]
# (N.B. pathlib.Path doesn't accept bytes so made dummy bytes pathlike for demo)
shutil.move(DummyPathLike(b"test2.txt"), DummyPathLike(b"test1.txt"))

# There's no constraints between the src and dst path types so no need for type vars.
shutil.move("test1.txt", DummyPathLike(b"test2.txt"))

Path("test2.txt").unlink()
```

mypy gives the following false positives:
```
type "bytes"; expected "Union[str, PathLike[str]]"
test.py:24: error: Argument 2 to "move" has incompatible type "bytes"; expected "Union[str, PathLike[str]]"
test.py:31: error: Argument 1 to "DummyPathLike" has incompatible type "bytes"; expected "str"
test.py:34: error: Argument 1 to "DummyPathLike" has incompatible type "bytes"; expected "str"
Found 4 errors in 1 file (checked 1 source file)
```

Should be a quick fix to switch StrPath with StrOrBytesPath in the stub so will pick this up now 👍

### PR DESCRIPTION
**Pull request by jpy-git**
Closes #6831. Thorough example included in the issue.

TLDR: `shutil.move` can take bytes arguments

### CODE REVIEW

File: stdlib/shutil.pyi [hunk 1]
```diff
@@ -49,11 +49,11 @@ def rmtree(path: bytes | StrPath, ignore_errors: bool = ..., onerror: Callable[[
 _CopyFn = Union[Callable[[str, str], None], Callable[[StrPath, StrPath], None]]
 
 if sys.version_info >= (3, 9):
-    def move(src: StrPath, dst: StrPath, copy_function: _CopyFn = ...) -> _PathReturn: ...
+    def move(src: StrOrBytesPath, dst: StrOrBytesPath, copy_function: _CopyFn = ...) -> _PathReturn: ...
 
 else:
     # See https://bugs.python.org/issue32689
-    def move(src: str, dst: StrPath, copy_function: _CopyFn = ...) -> _PathReturn: ...
+    def move(src: str | bytes, dst: StrOrBytesPath, copy_function: _CopyFn = ...) -> _PathReturn: ...
```
Comment (jpy-git) [inline stdlib/shutil.pyi:56]: FYI the linked bug applies to `Pathlike[bytes]` as well, but I've verified it works fine with `bytes` for Python < 3.9.

```python
>>> import shutil
>>> shutil.move(b"test1.txt", b"test2.txt")
'test2.txt'
```

**Comment by Akuli** [comment]
I don't think mixing str and bytes paths works. The source code has this:

```
real_dst = os.path.join(dst, _basename(src))
```
which fails if it tries to join strings and bytes.

**Comment by jpy-git** [comment]
> I don't think mixing str and bytes paths works. The source code has this:
> 
> ```
> real_dst = os.path.join(dst, _basename(src))
> ```
> 
> which fails if it tries to join strings and bytes.

This def works for me:
```python
>>> import shutil
>>> shutil.move("test.txt", b"test2.txt")
b'test2.txt'
>>> shutil.move(b"test2.txt", "test.txt")
'test.txt'
```

**Comment by Akuli** [comment]
The code I mentioned runs when `dst` is a directory.

**Comment by jpy-git** [comment]
> The code I mentioned runs when `dst` is a directory.

Ahh it's only when `dst` is an *existing* directory.

In that case yes I get errors:
```
>>> shutil.move("some_folder", b"some_existing_folder")
...
TypeError: Can't mix strings and bytes in path components
>>> shutil.move(b"some_folder", "some_exising_folder")
...
File "/usr/local/Cellar/python@3.9/3.9.9/Frameworks/Python.framework/Versions/3.9/lib/python3.9/shutil.py", line 771, in _basename
    return os.path.basename(path.rstrip(sep))
TypeError: a bytes-like object is required, not 'str'
```

**Comment by jpy-git** [comment]
The second of those issues seems like a bigger issue since this would also fail:
```
>>> shutil.move(b"some_folder", b"some_exising_folder")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/Cellar/python@3.9/3.9.9/Frameworks/Python.framework/Versions/3.9/lib/python3.9/shutil.py", line 810, in move
    real_dst = os.path.join(dst, _basename(src))
  File "/usr/local/Cellar/python@3.9/3.9.9/Frameworks/Python.framework/Versions/3.9/lib/python3.9/shutil.py", line 771, in _basename
    return os.path.basename(path.rstrip(sep))
TypeError: a bytes-like object is required, not 'str'
```

**Comment by jpy-git** [comment]
With those test cases, I think we should close this PR since bytes will always break the existing folder case.

Do you happen to know where the best place flag this issue is for cpython?

**Comment by jpy-git** [comment]
No worries, it looks like the types `shutil` functions accept is a bit inconsistent in general (although looks like they're improving) 😄 

I'm working through the other `shutil` functions atm as I'm finding a lot of them work fine with `bytes`. Once I've gone through all the others we should hopefully have quite a strong precedent for getting `move` updated to accept `bytes` arguments in the existing `dst` case.

In the meantime, maybe it's worth me removing my changes and just adding a comment above the `move` stub mentioning the reasoning for not accepting `bytes` argument (or a link to this PR) so people are aware of our reasoning for not changing it (given that from my examples its easy to be misled by this function)?

**Comment by JelleZijlstra** [comment]
> In the meantime, maybe it's worth me removing my changes and just adding a comment above the move stub mentioning the reasoning for not accepting bytes argument (or a link to this PR) so people are aware of our reasoning for not changing it (given that from my examples its easy to be misled by this function)?

That seems right to me.

**Comment by jpy-git** [comment]
I've reverted the changes and added a comment linking to this issue to highlight for future contributors 👍

**Review by Akuli — APPROVED** [review]
Thanks!

**Comment by Akuli** [comment]
> Do you happen to know where the best place flag this issue is for cpython?

`bugs.python.org`. My experience with reporting and fixing bugs in cpython hasn't been great, but you might have better luck.
